edgelet/iotedge/src/check/checks/identity_certificate_expiry.rs

```diff
@@ -0,0 +1,150 @@
+use std::fs::File;
+use std::io::Read;
+use std::path::PathBuf;
+
+use failure::{self, Context, ResultExt};
+
+use edgelet_core::{self, AttestationMethod, ManualAuthMethod, Provisioning, RuntimeSettings};
+
+use crate::check::{checker::Checker, Check, CheckResult};
+
+#[derive(Default, serde_derive::Serialize)]
+pub struct IdentityCertificateExpiry {
+    // provisioning: Option<Provisioning>, /* Note: contains shared keys */
+    certificate_info: Option<CertificateValidity>,
+}
+impl Checker for IdentityCertificateExpiry {
+    fn id(&self) -> &'static str {
+        "identity-certificate-expiry"
+    }
+    fn description(&self) -> &'static str {
+        "production readiness: identity certificates expiry"
+    }
+    fn result(&mut self, check: &mut Check) -> CheckResult {
+        self.execute(check).unwrap_or_else(CheckResult::Failed)
+    }
+    fn get_json(&self) -> serde_json::Value {
+        serde_json::to_value(self).unwrap()
+    }
+}
+impl IdentityCertificateExpiry {
+    fn execute(&mut self, check: &mut Check) -> Result<CheckResult, failure::Error> {
+        let settings = if let Some(settings) = &check.settings {
+            settings
+        } else {
+            return Ok(CheckResult::Skipped);
+        };
+        // self.provisioning = Some(settings.provisioning().to_owned());
+
+        match settings.provisioning() {
+            Provisioning::Dps(dps) => {
+                if let AttestationMethod::X509(x509_info) = dps.attestation() {
+                    let path = x509_info.identity_cert()?;
+
+                    let result =
+                        CertificateValidity::parse("DPS identity certificate".to_owned(), path)?;
+                    self.certificate_info = Some(result.clone());
+                    return result.to_check_result();
+                }
+            }
+            Provisioning::Manual(manual) => {
+                if let ManualAuthMethod::X509(x509) = manual.authentication_method() {
+                    let path = x509.identity_cert()?;
+                    let result = CertificateValidity::parse(
+                        "Manual authentication identity certificate".to_owned(),
+                        path,
+                    )?;
+                    self.certificate_info = Some(result.clone());
+                    return result.to_check_result();
+                }
+            }
+            Provisioning::External(_) => (),
+        }
+
+        Ok(CheckResult::Ignored)
+    }
+}
+
+#[derive(Debug, serde_derive::Serialize, Clone)]
+pub struct CertificateValidity {
+    cert_name: String,
+    cert_path: PathBuf,
+    pub not_after: chrono::DateTime<chrono::Utc>,
```

Comment:
This seems like it could be private.

Reply:
it's used in certificate_quickstart.rs

let CertificateValidity { not_after, .. } = CertificateValidity::parse(